docs/admin/etcd.md

```diff
@@ -1,49 +1,196 @@
+volatile…_
 
-By default, Kubernetes objects are stored under the `/registry` key in etcd.
-This path can be prefixed by using the [kube-apiserver](/docs/admin/kube-apiserver) flag
-`--etcd-prefix="/foo"`.
+## Design
 
-`etcd` is the only place that Kubernetes keeps state.
+This section describes how we are going to do the migration given the [etcd upgrade
+requirements](#etcd-upgrade-requirements).
 
-## Troubleshooting
+Note that since the code changes in Kubernetes code needed
+to support etcd v3 API are pretty local and not very sophisticated, we will not
+focus on it at all. We are only focusing on the upgrade/rollback here.
 
-To test whether `etcd` is running correctly, you can try writing a value to a
-test key. On your master VM (or somewhere with firewalls configured such that
-you can talk to your cluster's etcd), try:
+### New etcd docker image
+We decided to completely change the way how etcd image looks like (and how it
+works). So far the docker image for etcd in version X contained only etcd and
+etcdctl binaries in version X.
 
-```shell
-curl -fs -X PUT "http://${host}:${port}/v2/keys/_test"
+Going forward, the docker image for etcd in version X will contain multiple
+versions of etcd. For example 3.0.17 image will contain all: 2.2.1, 2.3.7 and
+3.0.17 binaries of etcd and etcdctl. This will allow running etcd in multiple
+different versions using exactly the same docker image.
+
+Additionally, the image will contain our custom script for doing migration
+between versions (written by Kubernetes team) and the custom rollback tool
+(provided by the etcd team).
+
+### Migration script
+The migration script that will now be part of etcd docker image is a bash
+script, working as following:
+
+1. Detect which version of etcd we were previously running.
+   For that purpose, we have added a dedicated file (`version.txt`) that
+   holds that information and is stored in the etcd-data-specific directory
+   (next to the etcd data). If the file doesn’t exist, we default it to 2.2.1
+   version.
+1. If we are in 2.2.1 version and are supposed to upgrade, backup
+   data.
+1. Based on the detected previous etcd version and the desired one (the desired
+   version is communicated via environment variable), do the upgrade steps (if
+   needed).
+   That means, for every minor etcd release between the detected one (excluding)
+   to the desired one (including):
+   1. start etcd in that version
+   1. wait until it is healthy (healthy means that you can write some data to
+      it)
+   1. stop this etcd
+   The important thing here is that those etcd will not listen on the default
+   etcd port.  They are hardcoded to listen on ports which the apiserver is not
+   configured to connect to, which means that apiserver won’t be able to connect
+   to those etcd's.  Assuming no other client goes out of its way to try to
+   connect and write to this obscure port, no new data will be written during
```

Comment:
I remain unsure of what the procedure is for an HA cluster, trying to tease it out in https://github.com/kubernetes/kubernetes/issues/43364

Because we don't change the peer port AFAIK, other apiservers will write to their local etcds and these writes _will_ be replicated to the node being upgraded.  Even if not, in a 3 node cluster, 2 of the nodes will make progress while the first is being migrated.

It sounds like the procedure is to either put the apiserver into read-only mode or shut down the etcd cluster entirely (unsure as to how we're doing this in our e2e tests of this scenario - if anyone knows I'd like to follow the same procedure that we're testing).

Hopefully the upgrade works when the members were not in sync as well.